clients/src/main/java/org/apache/kafka/common/requests/StopReplicaRequest.java

```diff
@@ -17,149 +17,153 @@
+        super(ApiKeys.STOP_REPLICA, version);
+        this.data = data;
     }
 
     public StopReplicaRequest(Struct struct, short version) {
-        super(ApiKeys.STOP_REPLICA, struct, version);
-
-        partitions = new HashSet<>();
-        if (version > 0) { // V1
-            for (Object topicObj : struct.get(PARTITIONS)) {
-                Struct topicData = (Struct) topicObj;
-                String topic = topicData.get(TOPIC_NAME);
-                for (Object partitionObj : topicData.get(PARTITION_IDS)) {
-                    int partition = (Integer) partitionObj;
-                    partitions.add(new TopicPartition(topic, partition));
-                }
-            }
-        } else { // V0
-            for (Object partitionDataObj : struct.get(PARTITIONS)) {
-                Struct partitionData = (Struct) partitionDataObj;
-                String topic = partitionData.get(TOPIC_NAME);
-                int partition = partitionData.get(PARTITION_ID);
-                partitions.add(new TopicPartition(topic, partition));
-            }
-        }
-        deletePartitions = struct.get(DELETE_PARTITIONS);
+        this(new StopReplicaRequestData(struct, version), version);
     }
 
     @Override
     public StopReplicaResponse getErrorResponse(int throttleTimeMs, Throwable e) {
         Errors error = Errors.forException(e);
 
-        Map<TopicPartition, Errors> responses = new HashMap<>(partitions.size());
-        for (TopicPartition partition : partitions) {
-            responses.put(partition, error);
+        StopReplicaResponseData data = new StopReplicaResponseData();
+        data.setErrorCode(error.code());
+        List<StopReplicaPartitionError> partitions = new ArrayList<>();
+        for (TopicPartition tp : partitions()) {
+            partitions.add(new StopReplicaPartitionError()
+                .setTopicName(tp.topic())
+                .setPartitionIndex(tp.partition())
+                .setErrorCode(error.code()));
         }
+        data.setPartitionErrors(partitions);
 
         short versionId = version();
         switch (versionId) {
             case 0:
             case 1:
-                return new StopReplicaResponse(error, responses);
+                return new StopReplicaResponse(data);
             default:
                 throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
                         versionId, this.getClass().getSimpleName(), ApiKeys.STOP_REPLICA.latestVersion()));
         }
     }
 
     public boolean deletePartitions() {
-        return deletePartitions;
+        return data.deletePartitions();
+    }
+
+    /**
+     * Note that this method has allocation overhead per iterated element, so callers should copy the result into
+     * another collection if they need to iterate more than once.
+     */
+    // Implementation note: we should strive to avoid allocation overhead per element, see
```

Comment:
This is very nitpicky, I know, but I would really prefer to just extend the C-style comment than to mix C and C++ styles here.

Reply:
While we're nitpicking, the former is a javadoc comment. :) That's why I didn't include the implementation node in it. Since this is all internal anyway, I made the change you suggested.